build: start build on z/OS

alexcfyung opened the pull request:
Update configure.py, Makefile, and common.gypi to get the
build started on z/OS.

Just with this commit node.js won't build on z/OS as it just start the build. 
This will be the first pr to upstream z/OS changes.

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [ ] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [ ] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

<!--
Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

vtjnash committed:
tls: improve handling of shutdown

RFC 5246 section-7.2.1 requires that the implementation must immediately
stop reading from the stream, as it is no longer TLS-encrypted. The
underlying stream is permitted to still pump events (and errors) to
other users, but those are now unencrypted, so we should not process
them here. But therefore, we do not want to stop the underlying stream,
as there could be another user of it, but we do need to remove ourselves
as a listener.

Per TLS v1.2, we should have also destroy the TLS state entirely here
(including the writing side), but this was revised in TLS v1.3 to permit
the stream to continue to flush output.

There appears to be some inconsistencies in the way nodejs handles
ownership of the underlying stream, with `TLS.close()` on the write side
also calling shutdown on the underlying stream (thus assuming other
users of the underlying stream are not permitted), while receiving EOF
on the read side leaves the underlying channel open. These
inconsistencies are left for a later person to resolve, if the extra
functionality is needed (as described in #35904). The current goal here
is to the fix the occasional CI exceptions depending on the timing of
these kernel messages through the TCP stack.

PR-URL: https://github.com/nodejs/node/pull/36111
Fixes: https://github.com/nodejs/node/pull/35946
Refs: https://github.com/libuv/libuv/pull/3036
Refs: https://github.com/nodejs/node/issues/35904
Co-authored-by: Momtchil Momtchev <momtchil@momtchev.com>
Reviewed-By: Robert Nagy <ronagy@icloud.com>
Reviewed-By: Matteo Collina <matteo.collina@gmail.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>

gabylb committed:
build: start build on z/OS

Update configure.py, Makefile, and common.gypi to get the
build started on z/OS.

PiotrRybak committed:
doc: add note about pip being required

Clean install requires pip to make coverage but instructions didn't
mention it.

PR-URL: https://github.com/nodejs/node/pull/40669
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: James M Snell <jasnell@gmail.com>

Gena888 committed:
doc: add explicit declaration of fd with null val

In example of implementing a writable stream with extending on
Writable add explicit declaration of 'fd' (file descriptor)
variable with null value.
It will make this example more similar to readable stream's one.
And will make it easier to figure out in topic.

PR-URL: https://github.com/nodejs/node/pull/40704
Reviewed-By: Tobias Nießen <tniessen@tnie.de>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: Minwoo Jung <nodecorelab@gmail.com>
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Matteo Collina <matteo.collina@gmail.com>

ronag committed:
stream: allow readable to end early without error

PR-URL: https://github.com/nodejs/node/pull/40881
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Matteo Collina <matteo.collina@gmail.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>

jasnell committed:
lib: add AbortSignal.timeout

Refs: https://github.com/whatwg/dom/pull/1032
Signed-off-by: James M Snell <jasnell@gmail.com>

PR-URL: https://github.com/nodejs/node/pull/40899
Reviewed-By: Michaël Zasso <targos@protonmail.com>
Reviewed-By: Matteo Collina <matteo.collina@gmail.com>

jasnell committed:
timers: add experimental scheduler api

Adds experimental implementations of the yield and wait APIs being
explored at https://github.com/WICG/scheduling-apis.

When I asked the WHATWG folks about the possibility of standardizing the
[awaitable versions of setTimeout/setImmediate](https://github.com/whatwg/html/issues/7340)
that we have implemented in `timers/promises`, they pointed at the work
in progress scheduling APIs draft as they direction they'll be going.
While there is definitely a few thing in that draft that have
questionable utility to Node.js, the yield and wait APIs map cleanly to
the setImmediate and setTimeout we already have.

Signed-off-by: James M Snell <jasnell@gmail.com>

PR-URL: https://github.com/nodejs/node/pull/40909
Reviewed-By: Matteo Collina <matteo.collina@gmail.com>
Reviewed-By: Antoine du Hamel <duhamelantoine1995@gmail.com>
Reviewed-By: Benjamin Gruenbaum <benjamingr@gmail.com>
Reviewed-By: Darshan Sen <raisinten@gmail.com>

jasnell committed:
lib: add abortSignal.throwIfAborted()

Refs: https://github.com/whatwg/dom/pull/1034
Signed-off-by: James M Snell <jasnell@gmail.com>

PR-URL: https://github.com/nodejs/node/pull/40951
Reviewed-By: Benjamin Gruenbaum <benjamingr@gmail.com>

BethGriggs committed:
2021-12-01, Version 16.13.1 'Gallium' (LTS)

Notable changes:

- **deps**: upgrade npm to 8.1.2 (npm team)
  [#40643](https://github.com/nodejs/node/pull/40643)
- **deps**: update c-ares to 1.18.1 (Richard Lau)
  [#40660](https://github.com/nodejs/node/pull/40660)
- **doc**: add VoltrexMaster to collaborators (voltrexmaster)
  [#40566](https://github.com/nodejs/node/pull/40566)
- **lib**: fix regular expression to detect \`/\` and \`\\\`
  (Francesco Trotta) [#40325](https://github.com/nodejs/node/pull/40325)

PR-URL: https://github.com/nodejs/node/pull/40974

jasnell committed:
errors: add support for cause in aborterror

Signed-off-by: James M Snell <jasnell@gmail.com>

PR-URL: https://github.com/nodejs/node/pull/41008
Reviewed-By: Ruben Bridgewater <ruben@bridgewater.de>
Reviewed-By: Robert Nagy <ronagy@icloud.com>
Reviewed-By: Benjamin Gruenbaum <benjamingr@gmail.com>

jasnell committed:
timers: propagate signal.reason in awaitable timers

Signed-off-by: James M Snell <jasnell@gmail.com>

PR-URL: https://github.com/nodejs/node/pull/41008
Reviewed-By: Ruben Bridgewater <ruben@bridgewater.de>
Reviewed-By: Robert Nagy <ronagy@icloud.com>
Reviewed-By: Benjamin Gruenbaum <benjamingr@gmail.com>

jasnell committed:
readline: propagate signal.reason in awaitable question

Signed-off-by: James M Snell <jasnell@gmail.com>

PR-URL: https://github.com/nodejs/node/pull/41008
Reviewed-By: Ruben Bridgewater <ruben@bridgewater.de>
Reviewed-By: Robert Nagy <ronagy@icloud.com>
Reviewed-By: Benjamin Gruenbaum <benjamingr@gmail.com>

jasnell committed:
stream: use cause options in AbortError constructors

Signed-off-by: James M Snell <jasnell@gmail.com>

PR-URL: https://github.com/nodejs/node/pull/41008
Reviewed-By: Ruben Bridgewater <ruben@bridgewater.de>
Reviewed-By: Robert Nagy <ronagy@icloud.com>
Reviewed-By: Benjamin Gruenbaum <benjamingr@gmail.com>

jasnell committed:
http2: propagate abortsignal reason in new AbortError constructor

Signed-off-by: James M Snell <jasnell@gmail.com>

PR-URL: https://github.com/nodejs/node/pull/41008
Reviewed-By: Ruben Bridgewater <ruben@bridgewater.de>
Reviewed-By: Robert Nagy <ronagy@icloud.com>
Reviewed-By: Benjamin Gruenbaum <benjamingr@gmail.com>

jasnell committed:
fs: propagate abortsignal reason in new AbortSignal constructors

Signed-off-by: James M Snell <jasnell@gmail.com>

PR-URL: https://github.com/nodejs/node/pull/41008
Reviewed-By: Ruben Bridgewater <ruben@bridgewater.de>
Reviewed-By: Robert Nagy <ronagy@icloud.com>
Reviewed-By: Benjamin Gruenbaum <benjamingr@gmail.com>

jasnell committed:
lib: propagate abortsignal reason in new AbortError constructor in blob

Signed-off-by: James M Snell <jasnell@gmail.com>

PR-URL: https://github.com/nodejs/node/pull/41008
Reviewed-By: Ruben Bridgewater <ruben@bridgewater.de>
Reviewed-By: Robert Nagy <ronagy@icloud.com>
Reviewed-By: Benjamin Gruenbaum <benjamingr@gmail.com>

jasnell committed:
events: propagate abortsignal reason in new AbortError ctor in events

Signed-off-by: James M Snell <jasnell@gmail.com>

PR-URL: https://github.com/nodejs/node/pull/41008
Reviewed-By: Ruben Bridgewater <ruben@bridgewater.de>
Reviewed-By: Robert Nagy <ronagy@icloud.com>
Reviewed-By: Benjamin Gruenbaum <benjamingr@gmail.com>

targos committed:
build: re-enable V8 concurrent marking

It was unintentionally disabled during a V8 update.

Fixes: https://github.com/nodejs/node/issues/41012

PR-URL: https://github.com/nodejs/node/pull/41013
Reviewed-By: Benjamin Gruenbaum <benjamingr@gmail.com>
Reviewed-By: Richard Lau <rlau@redhat.com>
Reviewed-By: Jiawen Geng <technicalcute@gmail.com>
Reviewed-By: James M Snell <jasnell@gmail.com>

jasnell committed:
events: propagate weak option for kNewListener

Signed-off-by: James M Snell <jasnell@gmail.com>

PR-URL: https://github.com/nodejs/node/pull/40899
Reviewed-By: Michaël Zasso <targos@protonmail.com>
Reviewed-By: Matteo Collina <matteo.collina@gmail.com>

mhdawson committed:
node-api,doc: document parms which can be optional

Fixes: https://github.com/nodejs/abi-stable-node/issues/252

I've been through the Node-api methods and I believe
this is the last places where optional parameters
were not documented.

Signed-off-by: Michael Dawson <mdawson@devrus.com>

PR-URL: https://github.com/nodejs/node/pull/41021
Reviewed-By: Mohammed Keyvanzadeh <mohammadkeyvanzade94@gmail.com>
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: Tobias Nießen <tniessen@tnie.de>

mhdawson committed:
inspector: add missing initialization

Add missing initialization flagged by Coverity

Signed-off-by: Michael Dawson <mdawson@devrus.com>

PR-URL: https://github.com/nodejs/node/pull/41022
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Richard Lau <rlau@redhat.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>

Trott committed:
tools: include JSDoc in ESLint updating tool

PR-URL: https://github.com/nodejs/node/pull/41027
Reviewed-By: Michaël Zasso <targos@protonmail.com>
Reviewed-By: Antoine du Hamel <duhamelantoine1995@gmail.com>

Trott committed:
tools: install and enable JSDoc linting in ESLint

PR-URL: https://github.com/nodejs/node/pull/41027
Reviewed-By: Michaël Zasso <targos@protonmail.com>
Reviewed-By: Antoine du Hamel <duhamelantoine1995@gmail.com>

Trott committed:
build: fail early in test-macos.yml

PR-URL: https://github.com/nodejs/node/pull/41035
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Michaël Zasso <targos@protonmail.com>

Trott committed:
tools: udpate packages in tools/doc

PR-URL: https://github.com/nodejs/node/pull/41036
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: Yash Ladha <yash@yashladha.in>

Trott committed:
build: add tools/doc to tools.yml updates

PR-URL: https://github.com/nodejs/node/pull/41036
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: Yash Ladha <yash@yashladha.in>

joyeecheung committed:
src: use a higher limit in the NearHeapLimitCallback

V8 requires the NearHeapLimitCallback to return a limit that's higher
than the initial one or otherwise it will crash.

PR-URL: https://github.com/nodejs/node/pull/41041
Refs: https://github.com/nodejs/node/pull/41013
Reviewed-By: Michaël Zasso <targos@protonmail.com>
Reviewed-By: Tobias Nießen <tniessen@tnie.de>
Reviewed-By: Anna Henningsen <anna@addaleax.net>
Reviewed-By: James M Snell <jasnell@gmail.com>

ronag committed:
stream: cleanup eos

PR-URL: https://github.com/nodejs/node/pull/40998
Reviewed-By: Ruben Bridgewater <ruben@bridgewater.de>
Reviewed-By: Benjamin Gruenbaum <benjamingr@gmail.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: Matteo Collina <matteo.collina@gmail.com>
Reviewed-By: James M Snell <jasnell@gmail.com>

BridgeAR committed:
readline: skip escaping characters again

This is a minor performance improvement for readline. It skips to
escape individual characters again after escaping them before.

Signed-off-by: Ruben Bridgewater <ruben@bridgewater.de>

PR-URL: https://github.com/nodejs/node/pull/41005
Reviewed-By: Michaël Zasso <targos@protonmail.com>
Reviewed-By: Benjamin Gruenbaum <benjamingr@gmail.com>
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Anna Henningsen <anna@addaleax.net>

jasnell committed:
stream: remove whatwg streams experimental warning

The API is still experimental, but the warning isn't necessary any
longer

Signed-off-by: James M Snell <jasnell@gmail.com>

PR-URL: https://github.com/nodejs/node/pull/40971
Refs: https://github.com/nodejs/node/issues/40950
Reviewed-By: Benjamin Gruenbaum <benjamingr@gmail.com>
Reviewed-By: Gerhard Stöbich <deb2001-github@yahoo.de>
Reviewed-By: Robert Nagy <ronagy@icloud.com>

jasnell committed:
lib: make AbortSignal cloneable/transferable

Allows for using `AbortSignal` across worker threads and contexts.

```js
const ac = new AbortController();
const mc = new MessageChannel();
mc.port1.onmessage = ({ data }) => {
  data.addEventListener('abort', () => {
    console.log('aborted!');
  });
};
mc.port2.postMessage(ac.signal, [ac.signal]);
```

Signed-off-by: James M Snell <jasnell@gmail.com>

PR-URL: https://github.com/nodejs/node/pull/41050
Refs: https://github.com/whatwg/dom/issues/948
Reviewed-By: Anna Henningsen <anna@addaleax.net>
Reviewed-By: Robert Nagy <ronagy@icloud.com>

lpinca committed:
doc: specify that `message.socket` can be nulled

The `socket` property of the `IncomingMessage` object is nulled on the
server after calling `message.destroy()` and on the client after a
request completes and the socket is kept alive.

Fixes: https://github.com/nodejs/node/issues/41011

PR-URL: https://github.com/nodejs/node/pull/41014
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Robert Nagy <ronagy@icloud.com>

joransiu commented on common.gypi:
```diff
@@ -566,6 +573,42 @@
+          '-Wa,GOFF',
+          '-qARCH=10',
+          '-qASCII',
+          '-qTUNE=10',
```
Do you want to consider TUNE to a newer hardware level?

alexcfyung replied:
we thought about to move the ARCH and TUNE to 11 but decided to keep it as 10 as V2R4 still support EC12

joransiu replied:
For Java, we set ARCH to min hardware support, but TUNE to a more recent hardware that we expect most users to be on when they use the product.   Not sure how much of a difference that makes tho.

alexcfyung replied:
I build with just setting TUNE to 11 and notice not much performance difference (slight increase in acmeair and decrease in octane), do you think we should change TUNE to 11? @IgorTodorovskiIBM

IgorTodorovskiIBM replied:
How large was the perf increase in acmeair?  Also, is it reproducible?  Curious if TUNE(12) gives better results

alexcfyung replied:
The increase is ~2% and seems to be consistent after a couple run, I also tried TUNE(12) but the result is the same (~2% improvement)

alexcfyung replied:
I change it to 12

joransiu commented on common.gypi:
```diff
@@ -200,7 +204,10 @@
           }],
-          ['OS!="mac" and OS!="win"', {
+          ['OS=="zos"', {
+            'cflags': [ '-qINLINE=::150:100000' ]
```
Should add a comment to explain why we need this custom -qINLINE option.

BridgeAR committed:
util: always visualize cause property in errors during inspection

While inspecting errors, always visualize the cause. That property
is non-enumerable by default while being useful in general for
debugging.

Duplicated stack frames are hidden.

Signed-off-by: Ruben Bridgewater <ruben@bridgewater.de>

PR-URL: https://github.com/nodejs/node/pull/41002
Fixes: https://github.com/nodejs/node/issues/40859
Fixes: https://github.com/nodejs/node/issues/38725
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Michaël Zasso <targos@protonmail.com>

BridgeAR committed:
lib: do not lazy load EOL in blob

This addresses a comment by loading the EOL from the constants file.

Signed-off-by: Ruben Bridgewater <ruben@bridgewater.de>

PR-URL: https://github.com/nodejs/node/pull/41004
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Antoine du Hamel <duhamelantoine1995@gmail.com>
Reviewed-By: Mohammed Keyvanzadeh <mohammadkeyvanzade94@gmail.com>
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>
Reviewed-By: Tobias Nießen <tniessen@tnie.de>
Reviewed-By: Benjamin Gruenbaum <benjamingr@gmail.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>

GeoffreyBooth committed:
doc: link to commit queue guide

PR-URL: https://github.com/nodejs/node/pull/41030
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Antoine du Hamel <duhamelantoine1995@gmail.com>
Reviewed-By: Ruben Bridgewater <ruben@bridgewater.de>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: Richard Lau <rlau@redhat.com>

Trott committed:
tools: use jsdoc recommended rules

Enable all recommended JSDoc linting rules and disable the ones we don't
(yet?) meet. They can be enabled one by one by removing the lines that
turn them off.

This requires adding --max-warnings to the ESLint invocations in
Makefile and vcbuild.bat because the preset enables the recommended
rules as warnings and not errors.

PR-URL: https://github.com/nodejs/node/pull/41057
Reviewed-By: Michaël Zasso <targos@protonmail.com>
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Ruben Bridgewater <ruben@bridgewater.de>

Trott committed:
test: add missing JSDoc parameter name

PR-URL: https://github.com/nodejs/node/pull/41057
Reviewed-By: Michaël Zasso <targos@protonmail.com>
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Ruben Bridgewater <ruben@bridgewater.de>

Trott committed:
tools: enable jsdoc/require-param-name lint rule

PR-URL: https://github.com/nodejs/node/pull/41057
Reviewed-By: Michaël Zasso <targos@protonmail.com>
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Ruben Bridgewater <ruben@bridgewater.de>

BridgeAR committed:
assert,util: compare RegExp.lastIndex while using deep equal checks

Compare the `lastIndex` property of regular expressions next to the
flags and source property.

Fixes: https://github.com/nodejs/node/issues/28766

Signed-off-by: Ruben Bridgewater <ruben@bridgewater.de>

PR-URL: https://github.com/nodejs/node/pull/41020
Reviewed-By: Michaël Zasso <targos@protonmail.com>
Reviewed-By: Richard Lau <rlau@redhat.com>
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Tobias Nießen <tniessen@tnie.de>
Reviewed-By: Anto Aravinth <anto.aravinth.cse@gmail.com>

Commit:
test: add auth option case for url.format

PR-URL: https://github.com/nodejs/node/pull/40516
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Mohammed Keyvanzadeh <mohammadkeyvanzade94@gmail.com>
Reviewed-By: Qingyu Deng <i@ayase-lab.com>

mhdawson commented on configure.py:
```diff
@@ -1198,6 +1228,8 @@ def configure_node(o):
   elif target_arch in ('mips', 'mipsel', 'mips64el'):
     configure_mips(o, target_arch)
+  elif sys.platform == 'zos':
+    configure_static_zoslib(o)
```
Should this be `configure_zos` to be consistent with the other platforms. It also would make sense if we ever need more configuration for zos and add it to the function.

mhdawson commented on configure.py:
```diff
@@ -68,6 +69,9 @@
     "Flags that allows you to control whether you want to build against "
     "built-in dependencies or its shared representations. If necessary, "
     "provide multiple libraries with comma.")
+static_optgroup = parser.add_argument_group("Static libraries",
```
In this case for z/OS we always build against the additional library right?  Is the issue that it won't be in a know location within the OS?

gabylb replied:
In the upstream V8, zoslib is a dependency that will be downloaded (by `gclient sync`) and placed in `v8/third_party/zoslib`. There, I also build zoslib (auto, from `v8/BUILD.gn`) as a shared library (and all V8 executables link with its sidedeck `libzoslib.x`). Should we build Node.js based on zoslib being in `deps/v8/third_party/zoslib` and also access it as shared library instead of static as currently done?

IgorTodorovskiIBM replied:
One consideration is that we have some C lib function overrides that will only work when zoslib is linked statically: https://github.ibm.com/open-z/zoslib/blob/master/src/zos.cc#L678

But there is a way to workaround this.

mhdawson replied:
> Should we build Node.js based on zoslib being in deps/v8/third_party/zoslib and also access it as shared library instead 
of static as currently done?

My answer to the first part would be yes, but for the second part it is more consistent to link in dependencies statically so that you don't need to install any other libraries for the node.js executable to work.  If we linked dynamically there would then be the question of how you install the shared library into the OS and will Node.js find it?

mhdawson replied:
I see @IgorTodorovskiIBM comment also points to linking statically. 

I might revise my comment on the first part.  We don't build Node.js with gclient sync.

I think we should be linking statically, but the question would be if it should be considered an OS dependency. ie you need to install it into your z/OS env to be able to build, and then assume it has been installed into a standard location?

mhdawson replied:
For building in the community we would then need to makes sure that the community ansiblescripts installed into the right place.

gabylb replied:
> ie you need to install it into your z/OS env to be able to build, and then assume it has been installed into a standard location?

How about that standard location be `deps/v8/third_party/zoslib` (even if for Node.js we'd manually clone it into that location)? In Node.js, only `libv8_libbase.a` contains `libzoslib.a`, and `node` binary (no longer) links with `libzoslib.a` since it (through `libv8_libbase.a`) is already in `libnode.so`. 

> I think we should be linking statically

fyi - we currently link Node.js statically, but also ship `libzoslib.so` under `install/lib`, and updated the user's `.nodeenv` to set `LIBPATH`; that's because in `deps/v8/tools/testrunner/local/utils.py`, `IsS390SimdSupported()` uses an API from zoslib, so it loads `libzoslib.so` to call that API.

alexcfyung replied:
zlib have a similar problem that both v8 and nodejs requires it, the current way is that both v8 and nodejs have and use its own zlib, but it could be configured that v8 and node.js both use the node.js's zlib. Perhaps we can put a mirrored zoslib under `deps` which nodejs's v8 will also use, but since zoslib only required by z/OS, I don't know would it be a welcome change to the community.

mhdawson replied:
I may not understand libzoslib.so well enough and related issues but the model I had in mind was 

1) install os lib, it goes into standard location for os like /usr/lib
2) Link against library in that standard location.  

I don't think we want the complication of suggesting a new dep under the node.js ../deps tree if at all possible.

Maybe we can talk it through in our meeting on Monday.

mhdawson replied:
We discussed in the V8 meeting today and seems like we should just go with what we have in the PR

IgorTodorovskiIBM commented on common.gypi:
```diff
@@ -566,6 +573,42 @@
       }],
+      ['OS == "zos"', {
+        'defines': [
+          '_XOPEN_SOURCE_EXTENDED',
```
We should probably match the macros that we use in our wrapper:
https://github.ibm.com/open-z/node-build/blob/v16.x.zos/bin/njsc

For example, _ENHANCED_ASCII_EXT=0xFFFFFFFF should be set

alexcfyung replied:
I added the missing macros from the wrapper

IgorTodorovskiIBM commented on configure.py:
```diff
@@ -849,16 +860,32 @@ def pkg_config(pkg):
-  try:
-    proc = subprocess.Popen(shlex.split(cc) + ['-E', '-P', '-x', lang, '-'],
+  if sys.platform == 'zos':
+    f = tempfile.NamedTemporaryFile(delete=False, suffix=".cc")
```
I think this and the below code is assuming njsc as the build compiler.  Should we try with Wyvern (clang) and see if it's still required?  Or add a check based on the `cc`

alexcfyung replied:
I tried running `configure.py` without this change with njsc and it works, do you remember the reason of this code? were it suppose to create an error without it?

IgorTodorovskiIBM replied:
I think the v8 version of njsc had an issue with the generated code.  Perhaps they fixed it in later versions of njsc.

alexcfyung replied:
I will remove this code as it build fine without it

zsw007 commented on common.gypi:
```diff
@@ -566,6 +573,42 @@
+          '-qASM',
+        ],
+        'cflags_cc': [
+          '-qxclang=-std=c++1y',
```
```suggestion
          '-qxclang=-std=c++14',
```

As c++1y is deprecated.

Trott committed:
url: detect hostname more reliably in url.parse()

Based on existing tests and code comments, url.parse() is expected to
treat any URL containing user@host as having a hostname. However, it
turns out this behavior relies on the URL having a hash which is
surprising, to put it mildly. Detect the host even without the hash.

PR-URL: https://github.com/nodejs/node/pull/41031
Reviewed-By: Ruben Bridgewater <ruben@bridgewater.de>
Reviewed-By: Myles Borins <myles.borins@gmail.com>

Trott committed:
doc: expand entries for isIP(), isIPv4(), and isIPv6()

Closes: https://github.com/nodejs/node/issues/40966

PR-URL: https://github.com/nodejs/node/pull/41028
Fixes: https://github.com/nodejs/node/issues/40966
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: Myles Borins <myles.borins@gmail.com>
Reviewed-By: Brian White <mscdex@mscdex.net>

npm-robot committed:
deps: upgrade npm to 8.2.0

PR-URL: https://github.com/nodejs/node/pull/41065
Reviewed-By: Myles Borins <myles.borins@gmail.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: Rich Trott <rtrott@gmail.com>

alexcfyung committed:
build: use c++14 as c++1y is deprecated

Co-authored-by: Wayne Zhang <zsw007@gmail.com>

alexcfyung committed:
build: add comment for qINLINE option

alexcfyung committed:
build: refactor configure_zos

alexcfyung committed:
build: add missing macros

BridgeAR committed:
util: add numericSeparator to util.inspect

This adds the `numericSeparator` option to util.inspect. Using it
separates numbers by thousands adding the underscore accordingly.

Signed-off-by: Ruben Bridgewater <ruben@bridgewater.de>

PR-URL: https://github.com/nodejs/node/pull/41003
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Michaël Zasso <targos@protonmail.com>
Reviewed-By: Rich Trott <rtrott@gmail.com>

Trott committed:
tools: remove Babel from license-builder.sh

Babel is now installed as a dependency in our ESLint installation. The
current license-builder.sh will fail to find the license for Babel.

The license is sill present in the repository/code base/file system. So
as with all other dependencies that are installed in
tools/node_modules/eslint/node_modules and have an MIT license, we are
still complying with the license. The inclusion in our own LICENSE was
to clearly signal that things under tools/node_modules/@babel were not
covered by the Node.js license. Because it is now under
tools/node_modules/eslint, that "we aren't the license-holders of this
code" aspect of things is already covered by the inclusion of the ESLint
license.

PR-URL: https://github.com/nodejs/node/pull/41049
Reviewed-By: Michaël Zasso <targos@protonmail.com>
Reviewed-By: Antoine du Hamel <duhamelantoine1995@gmail.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: Tobias Nießen <tniessen@tnie.de>
Reviewed-By: James M Snell <jasnell@gmail.com>

lpinca committed:
test: deflake test-trace-atomics-wait

Add possible outcome to the `expectedTimelines` array.

Fixes: https://github.com/nodejs/node/issues/41010

PR-URL: https://github.com/nodejs/node/pull/41018
Reviewed-By: Ruben Bridgewater <ruben@bridgewater.de>
Reviewed-By: Mohammed Keyvanzadeh <mohammadkeyvanzade94@gmail.com>
Reviewed-By: James M Snell <jasnell@gmail.com>

richardlau committed:
tools: rollback highlight.js

Recent upgrade of highlight.js has broken the docs build on the
withoutintl builds.

PR-URL: https://github.com/nodejs/node/pull/41078
Refs: https://github.com/nodejs/node/issues/41077
Refs: https://github.com/nodejs/node/pull/41036
Reviewed-By: Antoine du Hamel <duhamelantoine1995@gmail.com>
Reviewed-By: Rich Trott <rtrott@gmail.com>

nodejs-github-bot committed:
meta: update AUTHORS

PR-URL: https://github.com/nodejs/node/pull/41088
Reviewed-By: Rich Trott <rtrott@gmail.com>
Reviewed-By: Michaël Zasso <targos@protonmail.com>

Trott committed:
lib: use consistent types in JSDoc @returns

PR-URL: https://github.com/nodejs/node/pull/41089
Reviewed-By: Michaël Zasso <targos@protonmail.com>
Reviewed-By: Ruben Bridgewater <ruben@bridgewater.de>

Trott committed:
tools: enable jsdoc/require-returns-check lint rule

PR-URL: https://github.com/nodejs/node/pull/41089
Reviewed-By: Michaël Zasso <targos@protonmail.com>
Reviewed-By: Ruben Bridgewater <ruben@bridgewater.de>

jonahsnider committed:
build: add `--without-corepack`

PR-URL: https://github.com/nodejs/node/pull/41060
Reviewed-By: Richard Lau <rlau@redhat.com>
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Antoine du Hamel <duhamelantoine1995@gmail.com>
Reviewed-By: Ruben Bridgewater <ruben@bridgewater.de>

Trott committed:
build: skip documentation generation if no ICU

Closes: https://github.com/nodejs/node/issues/41077

PR-URL: https://github.com/nodejs/node/pull/41091
Fixes: https://github.com/nodejs/node/issues/41077
Reviewed-By: Richard Lau <rlau@redhat.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: Michaël Zasso <targos@protonmail.com>
Reviewed-By: Antoine du Hamel <duhamelantoine1995@gmail.com>

Trott committed:
tools: update highlight.js to 11.3.1

PR-URL: https://github.com/nodejs/node/pull/41091
Fixes: https://github.com/nodejs/node/issues/41077
Reviewed-By: Richard Lau <rlau@redhat.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: Michaël Zasso <targos@protonmail.com>
Reviewed-By: Antoine du Hamel <duhamelantoine1995@gmail.com>

lpinca committed:
tools: update ESLint to 8.4.0

PR-URL: https://github.com/nodejs/node/pull/41085
Reviewed-By: Tobias Nießen <tniessen@tnie.de>
Reviewed-By: Rich Trott <rtrott@gmail.com>
Reviewed-By: Mohammed Keyvanzadeh <mohammadkeyvanzade94@gmail.com>
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>
Reviewed-By: Ruben Bridgewater <ruben@bridgewater.de>
Reviewed-By: Derek Lewis <DerekNonGeneric@inf.is>

BridgeAR committed:
util: make sure error causes of any type may be inspected

An error cause may be of any type. Handle all of them, no matter
if they are an error or not.

Fixes: https://github.com/nodejs/node/issues/41096

Signed-off-by: Ruben Bridgewater <ruben@bridgewater.de>

PR-URL: https://github.com/nodejs/node/pull/41097
Reviewed-By: Michaël Zasso <targos@protonmail.com>
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Tobias Nießen <tniessen@tnie.de>
Reviewed-By: Anto Aravinth <anto.aravinth.cse@gmail.com>

cclauss committed:
doc: document support building with Python 3.10 on Windows

PR-URL: https://github.com/nodejs/node/pull/41098
Reviewed-By: Tobias Nießen <tniessen@tnie.de>
Reviewed-By: Michaël Zasso <targos@protonmail.com>
Reviewed-By: Richard Lau <rlau@redhat.com>
Reviewed-By: Rich Trott <rtrott@gmail.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: Derek Lewis <DerekNonGeneric@inf.is>

BridgeAR committed:
util: serialize falsy cause values while inspecting errors

Signed-off-by: Ruben Bridgewater <ruben@bridgewater.de>

PR-URL: https://github.com/nodejs/node/pull/41097
Fixes: https://github.com/nodejs/node/issues/41096
Reviewed-By: Michaël Zasso <targos@protonmail.com>
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Tobias Nießen <tniessen@tnie.de>
Reviewed-By: Anto Aravinth <anto.aravinth.cse@gmail.com>

bnb committed:
doc: add @bnb as a collaborator

Signed-off-by: Tierney Cyren <hello@bnb.im>

PR-URL: https://github.com/nodejs/node/pull/41100
Reviewed-By: Rich Trott <rtrott@gmail.com>
Reviewed-By: Richard Lau <rlau@redhat.com>
Reviewed-By: Michaël Zasso <targos@protonmail.com>
Reviewed-By: Stephen Belanger <admin@stephenbelanger.com>

BridgeAR committed:
util: escape lone surrogate code points using .inspect()

Unpaired surrogate code points have no representation in UTF8.
Therefore, such code points are just "random" output that is
unreadable. Instead, escape the code points similar to C0 and C1
control characters.

Refs: https://unicodebook.readthedocs.io/unicode_encodings.html#utf-16-surrogate-pairs

Signed-off-by: Ruben Bridgewater <ruben@bridgewater.de>

PR-URL: https://github.com/nodejs/node/pull/41001
Reviewed-By: James M Snell <jasnell@gmail.com>

BridgeAR committed:
assert: use stricter stack frame detection in .ifError()

This makes sure arbitrary stack traces are not handled as stack
frames.

Signed-off-by: Ruben Bridgewater <ruben@bridgewater.de>

PR-URL: https://github.com/nodejs/node/pull/41006
Reviewed-By: James M Snell <jasnell@gmail.com>

mhdawson reviewed (approved):
LGTM, good to go from my perspective.

Trott committed:
test: skip ESLint tests if no Intl

ESLint dependency now requires Intl because it uses regexp unicode
character properties.

Fixes: https://github.com/nodejs/node/issues/41102

PR-URL: https://github.com/nodejs/node/pull/41105
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: Michaël Zasso <targos@protonmail.com>
Reviewed-By: Derek Lewis <DerekNonGeneric@inf.is>
Reviewed-By: Richard Lau <rlau@redhat.com>
Reviewed-By: Ruben Bridgewater <ruben@bridgewater.de>

notroid5 committed:
doc: clarify escaping for ES modules

This should make it clear(er) how to escape special characters like `#`
and `?`.

Ref: https://github.com/nodejs/node/issues/41052

PR-URL: https://github.com/nodejs/node/pull/41074
Reviewed-By: Antoine du Hamel <duhamelantoine1995@gmail.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>

Trott committed:
doc: remove OpenJSF Slack nodejs from support doc

The OpenJSF Slack nodejs channel has a description that reads:

> 🚨 For Node.js application development questions, see:
>
> - Node Slackers: https://www.nodeslackers.com/
> - Node.js help repo: https://github.com/nodejs/help

Let's not send people with support questions there just to get
ignored or redirected elsewhere.

PR-URL: https://github.com/nodejs/node/pull/41108
Reviewed-By: Antoine du Hamel <duhamelantoine1995@gmail.com>
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>
Reviewed-By: Beth Griggs <bgriggs@redhat.com>
Reviewed-By: Michael Dawson <midawson@redhat.com>
Reviewed-By: Tobias Nießen <tniessen@tnie.de>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: Mohammed Keyvanzadeh <mohammadkeyvanzade94@gmail.com>

Trott committed:
doc,lib,tools: align multiline comments

These changes are in preparation for enabling a JSDoc lint rule.

PR-URL: https://github.com/nodejs/node/pull/41109
Reviewed-By: Antoine du Hamel <duhamelantoine1995@gmail.com>
Reviewed-By: Michaël Zasso <targos@protonmail.com>
Reviewed-By: Ruben Bridgewater <ruben@bridgewater.de>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>

Trott committed:
tools: enable JSDoc check-alignment lint rule

PR-URL: https://github.com/nodejs/node/pull/41109
Reviewed-By: Antoine du Hamel <duhamelantoine1995@gmail.com>
Reviewed-By: Michaël Zasso <targos@protonmail.com>
Reviewed-By: Ruben Bridgewater <ruben@bridgewater.de>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>

Trott committed:
tools: strip comments from lint-md rollup output

Refs: https://github.com/nodejs/node/pull/41081#issuecomment-986259228

PR-URL: https://github.com/nodejs/node/pull/41092
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: Tobias Nießen <tniessen@tnie.de>
Reviewed-By: Michaël Zasso <targos@protonmail.com>
Reviewed-By: Tierney Cyren <hello@bnb.im>

BethGriggs committed:
doc: simplify major release preparation

Changes/rational:

- Change the branch creation to two months prior from three months prior
  to the release.
- Do not cherry-pick commits within the last month - mirror the master
  branch up until 1 week before the release.
- Dropped 'semver major cut-off' and replace with 'Inform TSC of any
  majors which land within a month of the release'.

PR-URL: https://github.com/nodejs/node/pull/40816
Reviewed-By: Rich Trott <rtrott@gmail.com>
Reviewed-By: Matteo Collina <matteo.collina@gmail.com>
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>
Reviewed-By: Richard Lau <rlau@redhat.com>
Reviewed-By: Antoine du Hamel <duhamelantoine1995@gmail.com>
Reviewed-By: Danielle Adams <adamzdanielle@gmail.com>
Reviewed-By: Robert Nagy <ronagy@icloud.com>
Reviewed-By: Michael Dawson <midawson@redhat.com>
Reviewed-By: Tobias Nießen <tniessen@tnie.de>

npm-robot committed:
deps: upgrade npm to 8.3.0

PR-URL: https://github.com/nodejs/node/pull/41127
Reviewed-By: Rich Trott <rtrott@gmail.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>

Trott committed:
lib: include return types in JSDoc

Add types to @returns JSDoc annotations where the type is missing.

PR-URL: https://github.com/nodejs/node/pull/41130
Reviewed-By: Michaël Zasso <targos@protonmail.com>
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: Ruben Bridgewater <ruben@bridgewater.de>

Trott committed:
tools: enable jsdoc/require-returns-type ESLint rule

PR-URL: https://github.com/nodejs/node/pull/41130
Reviewed-By: Michaël Zasso <targos@protonmail.com>
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: Ruben Bridgewater <ruben@bridgewater.de>

fracsinus committed:
fs: fix `length` option being ignored during `read()`

Currently, `length` in an options object is ignored.

PR-URL: https://github.com/nodejs/node/pull/40906
Reviewed-By: Robert Nagy <ronagy@icloud.com>
Reviewed-By: Antoine du Hamel <duhamelantoine1995@gmail.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>

aduh95 committed:
fs: fix error codes for `fs.cp`

The context passed into this error must have `.code`, `.syscall` and
`.message`.

Fixes: https://github.com/nodejs/node/issues/41104

PR-URL: https://github.com/nodejs/node/pull/41106
Reviewed-By: Rich Trott <rtrott@gmail.com>
Reviewed-By: Evan Lucas <evanlucas@me.com>
Reviewed-By: Ruben Bridgewater <ruben@bridgewater.de>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>

Trott committed:
tools: update ESLint to 8.4.1

PR-URL: https://github.com/nodejs/node/pull/41114
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: Tobias Nießen <tniessen@tnie.de>
Reviewed-By: Mohammed Keyvanzadeh <mohammadkeyvanzade94@gmail.com>

jakub-g committed:
doc: fix comments in test-fs-watch.js

PR-URL: https://github.com/nodejs/node/pull/41046
Reviewed-By: Mohammed Keyvanzadeh <mohammadkeyvanzade94@gmail.com>
Reviewed-By: Antoine du Hamel <duhamelantoine1995@gmail.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: Anto Aravinth <anto.aravinth.cse@gmail.com>
Reviewed-By: Rich Trott <rtrott@gmail.com>
Reviewed-By: Tobias Nießen <tniessen@tnie.de>

santigimeno committed:
http: don't write empty data on req/res end()

When calling OutgoingMessage.end() with empty data argument, avoid
writing to the socket unless there's still pending data to be sent.

Fixes: https://github.com/nodejs/node/issues/41062

PR-URL: https://github.com/nodejs/node/pull/41116
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: Gerhard Stöbich <deb2001-github@yahoo.de>
Reviewed-By: James M Snell <jasnell@gmail.com>

mhdawson committed:
src: fix limit calculation

Coverity reported that the use of sizeof along with pointer
arithmetic was likely an error as the pointer arithmetic
would already be accounting for the size of what the
pointer points to.

Looking at the code that looked right but removing the
extra sizeOf caused tests to fail.

Looking more closely it seems like we were not allocating
a big enough buffer but the extra sizeof was allowing
us to convert even though it might have been corrupting
memory.

Signed-off-by: Michael Dawson <mdawson@devrus.com>

PR-URL: https://github.com/nodejs/node/pull/41026
Reviewed-By: Antoine du Hamel <duhamelantoine1995@gmail.com>

Gozala committed:
buffer: fix `Blob` constructor on various `TypedArray`s

Fixes: https://github.com/nodejs/node/issues/40705
    
PR-URL: https://github.com/nodejs/node/pull/40706
Reviewed-By: Robert Nagy <ronagy@icloud.com>
Reviewed-By: Antoine du Hamel <duhamelantoine1995@gmail.com>
Reviewed-By: Minwoo Jung <nodecorelab@gmail.com>
Reviewed-By: James M Snell <jasnell@gmail.com>

Trott committed:
doc: move style guide to findable location

Move the style guide to doc/README.md so people might find it. The
current location is easily overlooked, as is evidenced by
https://github.com/nodejs/node/pull/41025

PR-URL: https://github.com/nodejs/node/pull/41119
Reviewed-By: Michaël Zasso <targos@protonmail.com>
Reviewed-By: Derek Lewis <DerekNonGeneric@inf.is>
Reviewed-By: Mohammed Keyvanzadeh <mohammadkeyvanzade94@gmail.com>
Reviewed-By: James M Snell <jasnell@gmail.com>

Trott committed:
build: fix commit-queue-rebase functionality

For commit-queue-rebase only, commit-queue.sh uses `git push` which
requires the credentials from the git checkout. Remove
`persist-credentials: false` from this workflow.

PR-URL: https://github.com/nodejs/node/pull/41140
Reviewed-By: Antoine du Hamel <duhamelantoine1995@gmail.com>
Reviewed-By: Ruben Bridgewater <ruben@bridgewater.de>
Reviewed-By: Michaël Zasso <targos@protonmail.com>

BridgeAR committed:
util: pass through the inspect function to custom inspect functions

This allows to use more portable custom inspect functions.

Fixes: https://github.com/nodejs/node/issues/35956

Signed-off-by: Ruben Bridgewater <ruben@bridgewater.de>

PR-URL: https://github.com/nodejs/node/pull/41019
Reviewed-By: Michaël Zasso <targos@protonmail.com>

nodejs-github-bot committed:
meta: update AUTHORS

PR-URL: https://github.com/nodejs/node/pull/41144
Reviewed-By: Rich Trott <rtrott@gmail.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>

nodejs-github-bot committed:
tools: update lint-md-dependencies to rollup@2.61.1 vfile-reporter@7.0.3

PR-URL: https://github.com/nodejs/node/pull/41150
Reviewed-By: Rich Trott <rtrott@gmail.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>

mhdawson committed:
doc: update nodejs-sec managers

I'll submit a follow on PR to better document security
stewards along with their onboarding. For now just fix
the list so it's not out of date.

Signed-off-by: Michael Dawson <mdawson@devrus.com>

PR-URL: https://github.com/nodejs/node/pull/41128
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Tobias Nießen <tniessen@tnie.de>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: Rich Trott <rtrott@gmail.com>

nodejs-github-bot committed:
tools: update doc to remark-rehype@10.1.0


PR-URL: https://github.com/nodejs/node/pull/41149
Reviewed-By: Rich Trott <rtrott@gmail.com>
Reviewed-By: Antoine du Hamel <duhamelantoine1995@gmail.com>

aduh95 committed:
fs: accept URL as argument for `fs.rm` and `fs.rmSync`

PR-URL: https://github.com/nodejs/node/pull/41132
Reviewed-By: Michaël Zasso <targos@protonmail.com>
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Gerhard Stöbich <deb2001-github@yahoo.de>
Reviewed-By: Ruben Bridgewater <ruben@bridgewater.de>
Reviewed-By: Mohammed Keyvanzadeh <mohammadkeyvanzade94@gmail.com>
Reviewed-By: Matteo Collina <matteo.collina@gmail.com>

alexcfyung committed:
build: use TUNE(12) for z/OS

dygabo committed:
loader: return package format from defaultResolve if known

This is a proposed modification of defaultResolve to return the package
format in case it has been found during package resolution.
The format will be returned as described in the documentation:
https://nodejs.org/api/esm.html#resolvespecifier-context-defaultresolve
There is one new unit test as well:
test/es-module/test-esm-resolve-type.js

PR-URL: https://github.com/nodejs/node/pull/40980
Reviewed-By: Bradley Farias <bradley.meck@gmail.com>
Reviewed-By: Gerhard Stöbich <deb2001-github@yahoo.de>
Reviewed-By: Antoine du Hamel <duhamelantoine1995@gmail.com>

gabylb commented on Makefile:
```diff
@@ -10,6 +10,9 @@ TEST_CI_ARGS ?=
 STAGINGSERVER ?= node-www
 LOGLEVEL ?= silent
 OSTYPE := $(shell uname -s | tr '[:upper:]' '[:lower:]')
+ifeq ($(findstring OS/390,$OSTYPE),OS/390)
+OSTYPE ?= OS390
+endif
```
OSTYPE here will be all small-caps following the `tr` above.
```suggestion
ifeq ($(findstring os/390,$OSTYPE),os/390)
OSTYPE ?= os390
endif
```

gabylb commented on Makefile:
```diff
@@ -806,6 +809,9 @@ endif # ifeq ($(DISTTYPE),release)
 DISTTYPEDIR ?= $(DISTTYPE)
 RELEASE=$(shell sed -ne 's/\#define NODE_VERSION_IS_RELEASE \([01]\)/\1/p' src/node_version.h)
 PLATFORM=$(shell uname | tr '[:upper:]' '[:lower:]')
+ifeq ($(findstring OS/390,$PLATFORM),OS/390)
+PLATFORM ?= OS390
+endif
```
Same as https://github.com/ibmruntimes/node-zos/pull/12/files#r768152118

RaisinTen committed:
process: add `getActiveResourcesInfo()`

This is supposed to be a public alternative of the private APIs,
`process._getActiveResources()` and `process._getActiveHandles()`. When
called, it returns an array of strings containing the types of the
active resources that are currently keeping the event loop alive.

Signed-off-by: Darshan Sen <darshan.sen@postman.com>

PR-URL: https://github.com/nodejs/node/pull/40813
Reviewed-By: Stephen Belanger <admin@stephenbelanger.com>
Reviewed-By: Vladimir de Turckheim <vlad2t@hotmail.com>
Reviewed-By: Matteo Collina <matteo.collina@gmail.com>

RaisinTen committed:
assert: prefer reference comparison over string comparison

Pointer comparison takes constant time and string comparison takes
linear time unless there is some string interning going on, so this
might be a little bit faster.

Signed-off-by: Darshan Sen <darshan.sen@postman.com>

PR-URL: https://github.com/nodejs/node/pull/41015
Reviewed-By: Ruben Bridgewater <ruben@bridgewater.de>
Reviewed-By: Tobias Nießen <tniessen@tnie.de>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: Benjamin Gruenbaum <benjamingr@gmail.com>
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Zijian Liu <lxxyxzj@gmail.com>

silverwind committed:
meta: move silverwind to emeriti

PR-URL: https://github.com/nodejs/node/pull/41171
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>
Reviewed-By: Richard Lau <rlau@redhat.com>
Reviewed-By: Michaël Zasso <targos@protonmail.com>
Reviewed-By: Rich Trott <rtrott@gmail.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>

Ayase-252 committed:
doc: align module resolve algorithm with implementation


Fixes: https://github.com/nodejs/node/issues/38128

PR-URL: https://github.com/nodejs/node/pull/38837
Reviewed-By: Guy Bedford <guybedford@gmail.com>

Mesteery committed:
build: use gh cli in workflows file

PR-URL: https://github.com/nodejs/node/pull/40985
Reviewed-By: Antoine du Hamel <duhamelantoine1995@gmail.com>

Mesteery committed:
build: fix comment-labeled workflow

Refs: https://github.com/nodejs/node/pull/40985#issuecomment-994056146

PR-URL: https://github.com/nodejs/node/pull/41176
Reviewed-By: Antoine du Hamel <duhamelantoine1995@gmail.com>
Reviewed-By: Mohammed Keyvanzadeh <mohammadkeyvanzade94@gmail.com>
Reviewed-By: Rich Trott <rtrott@gmail.com>

alexcfyung committed:
build: change OSTYPE to lowercase for z/OS

Co-Authored-By: Gaby Baghdadi <baghdadi@ca.ibm.com>

alexcfyung committed:
build: remove unnecessary code when checking compiler

gabylb reviewed (approved):
LGTM

IgorTodorovskiIBM reviewed (approved):
LGTM

alexcfyung commented:
Closed because changes doesn't updated. Reopen at #15
